### PR TITLE
Make DH_check_pub_key() and DH_generate_key() safer yet

### PR DESCRIPTION
**Pull request by levitte**
We already check for an excessively large P in `DH_generate_key()`, but not in `DH_check_pub_key()`, and none of them check for an excessively large Q.

This change adds all the missing excessive size checks of P and Q.

Fixes CVE-2023-5678

### CODE REVIEW

**Comment by levitte** [comment]
These changes don't cherry-pick very well to release branches, so I'll make separate PRs for them

File: crypto/dh/dh_check.c [hunk 1]
```diff
@@ -155,7 +155,7 @@ int DH_check(const DH *dh, int *ret)
     /* Don't do any checks at all with an excessively large modulus */
     if (BN_num_bits(dh->params.p) > OPENSSL_DH_CHECK_MAX_MODULUS_BITS) {
         ERR_raise(ERR_LIB_DH, DH_R_MODULUS_TOO_LARGE);
-        *ret = DH_MODULUS_TOO_LARGE | DH_CHECK_P_NOT_PRIME;
+        *ret = DH_MODULUS_TOO_LARGE;
```
Comment (t8m) [inline crypto/dh/dh_check.c:158]: This is a safety setting, please keep it.

**Review by t8m — CHANGES_REQUESTED** [review]

**Comment by levitte** [comment]
This can cherry-pick to the 3.2 branch

**Review by mattcaswell — COMMENTED** [review]
Looks ok, pending the change requested by @t8m

File: crypto/dh/dh_check.c (hunk 1 above)
Comment (t8m) [reply crypto/dh/dh_check.c:158]: The return value from the function is 0. That means the *ret does not even pretend to be correct. The problem is not with the backwards compatibility with old versions but keeping applications that do some buggy checking of the *ret value which ignore the return value from the function call safe.

Comment (t8m) [reply crypto/dh/dh_check.c:158]: If we returned 1 here it would be a different thing but no, we don't.

Comment (levitte) [reply crypto/dh/dh_check.c:158]: ... er, are you arguing against setting `*ret` at all?  'cause by extension, you seem to be arguing a pointlessness regarding the specific flags, i.e. that what anyone might actually check is `ret != 0` if they check it at all.

Comment (levitte) [reply crypto/dh/dh_check.c:158]: Looking back at the PR where this was created, `DH_CHECK_P_NOT_PRIME` was argued as an *optional* addition.  I would have argued against it, had I looked.

Ref: https://github.com/openssl/openssl/pull/21524#issuecomment-1647788084

Comment (t8m) [reply crypto/dh/dh_check.c:158]: We would have then to agree to disagree and pass it over to OTC I am afraid.

**Comment by t8m** [comment]
**OTC: Do we prefer safety (for broken applications) over correctness (i.e., reverting https://github.com/openssl/openssl/commit/81d10e61a4b7d5394d08a718bf7d6bae20e818fc)?**

File: crypto/dh/dh_check.c (hunk 1 above)
Comment (t8m) [reply crypto/dh/dh_check.c:158]: Also strictly speaking this is a fairly unrelated change to this PR so if anything it should be in a separate PR.

Comment (levitte) [reply crypto/dh/dh_check.c:158]: You have a good point that this doesn't strictly belong in this PR.  I'll move it to another.

**Comment by levitte** [comment]
> **OTC: Do we prefer safety (for broken applications which check only DH_CHECK_P_NOT_PRIME bit) over correctness (i.e., partially reverting [81d10e6](https://github.com/openssl/openssl/commit/81d10e61a4b7d5394d08a718bf7d6bae20e818fc))?**

I'm moving the contending commit to another PR

**Review by t8m — APPROVED** [review]

**Review by mattcaswell — APPROVED** [review]

**Comment by hlandau** [comment]
Merged to master and 3.2. Thank you.

**Pull request closed without merging**